pkg/controller/build/buildrequest/machineosbuild.go

```diff
@@ -0,0 +1,250 @@
+package buildrequest
+
+import (
+	"context"
+	//nolint:gosec
+	"crypto/md5"
+	"fmt"
+
+	"github.com/distribution/reference"
+	"github.com/ghodss/yaml"
+	mcfgv1 "github.com/openshift/api/machineconfiguration/v1"
+	mcfgv1alpha1 "github.com/openshift/api/machineconfiguration/v1alpha1"
+	"github.com/openshift/machine-config-operator/pkg/controller/build/utils"
+	ctrlcommon "github.com/openshift/machine-config-operator/pkg/controller/common"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	clientset "k8s.io/client-go/kubernetes"
+)
+
+// This is the same salt / pattern from pkg/controller/render/hash.go
+var (
+	// salt is 80 random bytes.
+	// The salt was generated by `od -vAn -N80 -tu1 < /dev/urandom`. Do not change it.
+	salt = []byte{
+		16, 124, 206, 228, 139, 56, 175, 175, 79, 229, 134, 118, 157, 154, 211, 110,
+		25, 93, 47, 253, 172, 106, 37, 7, 174, 13, 160, 185, 110, 17, 87, 52,
+		219, 131, 12, 206, 218, 141, 116, 135, 188, 181, 192, 151, 233, 62, 126, 165,
+		64, 83, 179, 119, 15, 168, 208, 197, 146, 107, 58, 227, 133, 188, 238, 26,
+		33, 26, 235, 202, 32, 173, 31, 234, 41, 144, 148, 79, 6, 206, 23, 22,
+	}
+)
+
+// Holds the objects that are used to construct a MachineOSBuild with a hashed
+// name.
+type MachineOSBuildOpts struct {
+	MachineOSConfig   *mcfgv1alpha1.MachineOSConfig
+	MachineConfigPool *mcfgv1.MachineConfigPool
+	OSImageURLConfig  *ctrlcommon.OSImageURLConfig
+}
+
+// Validates that the required options are provided.
+func (m *MachineOSBuildOpts) validateForHash() error {
+	if m.MachineOSConfig == nil {
+		return fmt.Errorf("missing required MachineOSConfig")
+	}
+
+	if m.MachineConfigPool == nil {
+		return fmt.Errorf("missing required MachineConfigPool")
+	}
+
+	if m.MachineConfigPool.Name != m.MachineOSConfig.Spec.MachineConfigPool.Name {
+		return fmt.Errorf("name mismatch, MachineConfigPool has %q, MachineOSConfig has %q", m.MachineConfigPool.Name, m.MachineOSConfig.Spec.MachineConfigPool.Name)
+	}
+
+	if m.OSImageURLConfig == nil {
+		return fmt.Errorf("misssing OSImageURLConfig")
+	}
+
+	return nil
+}
+
+// Makes a deep-copy of the MachineOSConfig and clears the data from it that
+// can come from either the MachineOSConfig or from the OSImageURLConfig
+// object. We do this to stabilize the hashing of the name so that whether the
+// value comes from the MachineOSConfig or the OSImageURLConfig, the hash will
+// be same, provided that the value is the same.
+func (m *MachineOSBuildOpts) getMachineOSConfigForHashing() *mcfgv1alpha1.MachineOSConfig {
+	moscCopy := m.MachineOSConfig.DeepCopy()
+	moscCopy.Spec.BuildInputs.BaseOSImagePullspec = ""
+	moscCopy.Spec.BuildInputs.BaseOSExtensionsImagePullspec = ""
+	moscCopy.Spec.BuildInputs.ReleaseVersion = ""
+	return moscCopy
+}
+
+// Creates a list of objects that are consumed by the SHA256 hash.
```

Comment:
Note to reviewer: Does this seem like a good strategy? Basically, I'm trying to keep the hash somewhat stable where it should take the contents of the MachineOSConfig, the current rendered MachineConfig, and the base OS / extensions image / release version into consideration. For the base images and release version, there are actually two sources of truth for OCL: What is provided by the MachineOSConfig as well as what is provided by the OSImageURL ConfigMap (which I've implemented some code to formally extract and convert to a Go struct).

Reply:
non-blocking clarifying question:
What happens when a MOSC contains a valid BaseOSImagePullSpec / BaseOSExtensionsImagePullspec / Release Version and triggers a build. And then the OSImageURLConfig changes? 

Should the hashes change when BuildRequest.opts chooses MOSC pull secrets as source of truth over the OSImageURLConfig? In the current case it would cancel and rebuild the exact same image with just the change in hash